community/kernel/src/main/java/org/neo4j/kernel/impl/store/kvstore/AbstractKeyValueStore.java

```diff
@@ -0,0 +1,384 @@
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.neo4j.io.fs.FileSystemAbstraction;
+import org.neo4j.io.pagecache.PageCache;
+import org.neo4j.kernel.impl.locking.LockWrapper;
+import org.neo4j.kernel.lifecycle.LifecycleAdapter;
+
+import static org.neo4j.kernel.impl.locking.LockWrapper.readLock;
+import static org.neo4j.kernel.impl.locking.LockWrapper.writeLock;
+
+/**
+ * The base for building a key value store based on rotating immutable
+ * {@linkplain KeyValueStoreFile key/value store files}
+ *
+ * @param <Key>      a base type for the keys stored in this store.
+ * @param <MetaData> a type for containing the metadata of the store.
+ * @param <MetaDiff> a type that signifies changes in metadata on rotation.
+ */
+@Rotation(/*default strategy:*/Rotation.Strategy.LEFT_RIGHT/*(subclasses can override)*/)
+@State(/*default strategy:*/State.Strategy.CONCURRENT_HASH_MAP/*(subclasses can override)*/)
+public abstract class AbstractKeyValueStore<Key, MetaData, MetaDiff> extends LifecycleAdapter
+{
+    // TODO: add callbacks (for logging) on state transition events.
+    private final ReadWriteLock updateLock = new ReentrantReadWriteLock( /*fair=*/true );
+    private volatile KeyValueStoreState<Key, MetaData> state;
+    final int keySize, valueSize;
+
+    public AbstractKeyValueStore( FileSystemAbstraction fs, PageCache pages, File base, int keySize, int valueSize,
+                                  String... metadataKeys )
+    {
+        this.keySize = keySize;
+        this.valueSize = valueSize;
+        Rotation rotation = getClass().getAnnotation( Rotation.class );
+        Format format = new Format( metadataKeys );
+        this.state = getClass().getAnnotation( State.class ).value().initialState(
+                rotation.value().create( fs, pages, format, base, rotation.parameters() ), format );
+    }
+
+    @Override
+    public String toString()
+    {
+        return String.format( "%s[state=%s, hasChanges=%s]", getClass().getSimpleName(), state, state.hasChanges() );
+    }
+
+    protected final <Value> Value lookup( Key key, Reader<Value> lookup ) throws IOException
+    {
+        return state.lookup( key, lookup );
+    }
+
+    /** Introspective feature, not thread safe. */
+    protected final void visitAll( Visitor visitor ) throws IOException
+    {
+        try ( DataProvider provider = state.dataProvider() )
+        {
+            transfer( provider, visitor );
+        }
+    }
+
+    protected abstract Key readKey( ReadableBuffer key );
+
+    protected abstract void writeKey( Key key, WritableBuffer buffer );
+
+    protected abstract void writeFormatSpecifier( WritableBuffer formatSpecifier );
+
+    protected abstract Object parseHeader( String key, ReadableBuffer value );
+
+    protected abstract void extractMetadataEntry( MetaData metadata, String identifier, WritableBuffer value );
+
+    protected abstract MetaData initialMetadata();
```

Comment:
If we do what I said in https://github.com/neo4j/neo4j/pull/3852#discussion_r23917179 that would replace a lot of this logic. Although, `initialMetadata()` would still live here (the rest of the metadata methods would go away).